contracts/vortex/CarbonVortex.sol

```diff
@@ -0,0 +1,259 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.19;
+
+import { ReentrancyGuardUpgradeable } from "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import { Address } from "@openzeppelin/contracts/utils/Address.sol";
+
+import { ICarbonVortex } from "./interfaces/ICarbonVortex.sol";
+import { IVersioned } from "../utility/interfaces/IVersioned.sol";
+import { ICarbonController } from "../carbon/interfaces/ICarbonController.sol";
+import { Upgradeable } from "../utility/Upgradeable.sol";
+import { Token } from "../token/Token.sol";
+import { Utils } from "../utility/Utils.sol";
+import { MathEx } from "../utility/MathEx.sol";
+import { MAX_GAP, PPM_RESOLUTION } from "../utility/Constants.sol";
+
+interface IBancorNetwork {
+    function collectionByPool(Token pool) external view returns (address);
+
+    function tradeBySourceAmount(
+        Token sourceToken,
+        Token targetToken,
+        uint256 sourceAmount,
+        uint256 minReturnAmount,
+        uint256 deadline,
+        address beneficiary
+    ) external payable returns (uint256);
+}
+
+/**
+ * @dev CarbonVortex contract
+ */
+contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable, Utils {
+    using Address for address payable;
+
+    ICarbonController private immutable _carbonController;
+    IBancorNetwork private immutable _bancorNetwork;
+    Token private immutable _bnt;
+
+    uint256 private _totalBurned;
+
+    // rewards percentage
+    uint256 private _rewardsPPM;
```

Comment:
I do think it's a bit dangerous leaving it like this, since we might forget in the future that this is the result of porting `Rewards` to `_rewardsPPM`. What about keeping it as `uint32`, while declaring a `uint224 _reserved0` instead which can be set to `0` during `_postUpgrade` and deleted in a future upgrade?

Reply:
Alternatively, we can just deploy a new `CarbonVortex` instance (i.e., with `version` 1) and ignore the prior instance.

Reply:
Why is it dangerous ? 
We'll change the _rewardsPPM value during the upgrade, and this way it'll cover the entire slot. We'll have the proper `_rewardsPPM` value, so the rewards will be correct, and we won't have to think about this in future upgrades.

Reply:
Could you share why it worries you that we might forget it ?
The variable itself will be set in the `_postUpgrade` callback, meaning it would immediately become the proper value during upgrade.
Using `uint256` is cleaner in my opinion and since we'll set the value as the new `uint256`, we won't have issues with any dirty upper bits. Also it will cover the entire slot and won't corrupt the storage during a future upgrade.
If you mean to reserve the `uint224` slot in case we decide to use it in the future, it may make sense in this context.

Reply:
We discussed this offline. The TL;DR is that there are two options:
1. Keep it as is and add a `_postUpgrade` callback to this PR. 
2. Change it to `uint32`, so that the rest of the code will become similar to how `PPM`s are usually handled, while adding a reserved `uint224` (which will be also set in the `_postUpgrade` callback).

I'm totally fine with whatever @ivanzhelyazkov decides.